src/java/org/apache/cassandra/index/sai/disk/v1/SegmentBuilder.java

```diff
@@ -248,12 +283,56 @@ public long add(ByteBuffer term, PrimaryKey key, long sstableRowId)
 
         maxSegmentRowId = Math.max(maxSegmentRowId, segmentRowId);
 
-        long bytesAllocated = addInternal(term, segmentRowId);
+        long bytesAllocated = supportsAsyncAdd()
+                              ? addInternalAsync(term, segmentRowId)
+                              : addInternal(term, segmentRowId);
         totalBytesAllocated += bytesAllocated;
 
         return bytesAllocated;
     }
 
+    private long addInternalAsync(ByteBuffer term, int segmentRowId)
+    {
+        updatesInFlight.incrementAndGet();
+        ForkJoinPool.commonPool().submit(() -> {
+            try
+            {
+                long bytesAdded = addInternal(term, segmentRowId);
+                termSizeReservoir.update(bytesAdded);
+            }
+            catch (Throwable th)
+            {
+                asyncThrowable.compareAndExchange(null, th);
+            }
+            finally
+            {
+                updatesInFlight.decrementAndGet();
+            }
+        });
+        // bytes allocated will be approximated immediately as the average of recently added terms,
+        // rather than waiting until the async update completes to get the exact value.  The latter could
+        // result in a dangerously large discrepancy between the amount of memory actually consumed
+        // and the amount the limiter knows about if the queue depth grows.
```

Comment:
It is a little late to mention this, but if we find that the discrepancy is a problem, perhaps we could push the concurrent graph construction into the `GraphIndexBuilder` itself. I assume that `builder.addGraphNode(ordinal, vectorValues);` is the most expensive part of adding a row, and that step appears to have a constant cost per row within a `GraphIndexBuilder` instance. If we took that direction, the logic to write the graph would just wait for the in flight additions to complete.

I do not think it's worth the effort given this solution is already written.

Reply:
That's possible but since I imagine that most users have different ideas about the right way to control concurrency I'd prefer not to push those decisions into JVector.

Reply:
Works for me